public/js/components/Scopes.js

```diff
@@ -3,110 +3,169 @@
+  let scope = pauseInfo.getIn(["frame", "scope"]);
+  if (!scope) {
+    return null;
   }
 
-  return scopes;
-}
-
-function renderArgumentProperty(argumentBinding) {
-  const argumentName = argumentBinding.keySeq().first();
-  const actor = argumentBinding.first().getIn(["value", "actor"]);
-  return dom.li(
-    { className: "scope-property-list-item", key: actor },
-    argumentName
-  );
-}
-
-function renderVariableProperty(variableBinding, variableName) {
-  const actor = variableBinding.getIn(["value", "actor"]);
-  return dom.li(
-    { className: "scope-property-list-item", key: actor },
-    variableName
-  );
-}
-
-function renderScopeBindings(bindings) {
-  const variableBindings = bindings.get("variables");
-  const argumentBindings = bindings.get("arguments");
-
-  return dom.ul(
-    { className: "scope-property-list" },
-    argumentBindings.map(renderArgumentProperty),
-    variableBindings.map(renderVariableProperty)
-  );
-}
-
-function renderFunctionScope(scope, isFirst) {
-  const displayName = scope.getIn(["function", "displayName"]);
-  const functionName = displayName || "(anonymous)";
-  const label = "Function";
-
-  return {
-    header: `${label} [${functionName}]`,
-    component: () => renderScopeBindings(scope.get("bindings")),
-    opened: isFirst
-  };
-}
-
-/* TODO: render block with variables */
-function renderBlockScope(scope, isFirst) {
-  const label = "Block Scope";
-  return {
-    header: label,
-    component: () => renderScopeBindings(scope.get("bindings")),
-    opened: isFirst
-  };
-}
-
-function renderObjectScope(scope, isFirst) {
-  const objectClass = scope.getIn(["object", "class"]);
-  const label = objectClass;
+  const scopes = [];
+
+  do {
+    const type = scope.get("type");
+
+    if (type === "function" || type === "block") {
+      const bindings = scope.get("bindings");
+      const vars = bindings
+        .get("arguments").map(arg => arg.entrySeq().get(0))
+        .concat(bindings.get("variables").entrySeq())
+        .filter(binding => (!binding[1].hasIn(["value", "missingArguments"]) &&
+                            !binding[1].hasIn(["value", "optimizedOut"])))
+        .sort((a, b) => a[0].localeCompare(b[0]))
+        .toArray();
+
+      if (vars.length) {
+        let title;
+        if (type === "function") {
+          title = scope.getIn(["function", "displayName"]) || "(anonymous)";
+        } else {
+          title = "Block";
+        }
+
+        scopes.push([
+          title,
+          Map({
+            value: Map({
+              type: "object",
+              actor: Math.random().toString()
+            }),
+            prefetchedProperties: vars
+          }),
+        ]);
+      }
+    } else if (type === "object") {
+      scopes.push([
+        scope.getIn(["object", "class"]),
+        fromJS({
+          value: scope.get("object")
+        })
+      ]);
+    }
+  } while (scope = scope.get("parent")); // eslint-disable-line no-cond-assign
```

Comment:
Yeah, it's demo quality. We need to build the tree in a very different way so this will be totally re-worked when we have time to work on the object inspector again.